[Fix] Remove CastRemover for Tensor Indices (issue #386)

zzzDavid opened the pull request:
### Remove CastRemover for Tensor Indices 

**Fixed issue:** #386 

**Detailed description:** 

The cast operations on tensor indices are incorrectly removed by `CastRemover`, which causes binary operation data type mismatch. This PR fixes this issue by removing the CastRemovers and enforcing indices to have i32 data type for LLVM backend. 

**Link to the tests:** `tests/issues/test_issue_386.py`

chhzh123 commented on tvm/src/pass/generate_reuse_buffer.cc:
```diff
@@ -55,6 +55,99 @@ class ModulusRemover final : public IRMutator {
   std::map<const Variable*, Expr>& range_;
 };
 
+class CastRemover final : public IRMutator {
+ public:
+  CastRemover() {}
+
+  Expr Mutate_(const Cast* op, const Expr& e) {
+    return op->value;
+  }
+
+  Expr Mutate_(const Add* op, const Expr& e) {
+    Expr a = this->Mutate(op->a);
+    Expr b = this->Mutate(op->b);
+    if (const Cast* ca = a.as<Cast>()) {
+      a = ca->value;
+    }
+    if (const Cast* cb = b.as<Cast>()) {
+      b = cb->value;
+    }
+    return Add::make(a, b);
+  }
```
1. Is it safe to remove the cast operations?
2. Seems the `CastRemover` does not enumerate all the operations but only several binary ones.

zzzDavid replied:
I added some explanation below of why we do this, please see below the thread

chhzh123 reviewed (commented):
Can you also provide the test cases for `reuse_at`? Seems you changed the reuse buffer pass.

zzzDavid commented:
More explanation on why `CastRemover` is deleted in `tensor.py` and added in `generate_reuser_buffer.cc`:

The `ReuseBufferInserter` mutator calculates a tensor index expression to determine the reuse buffer size. The reuse buffer size needs to be a constant:

https://github.com/cornell-zhang/heterocl/blob/267d670bfd7ec14276a7ddd21103a6e6ed668e5b/tvm/src/pass/generate_reuse_buffer.cc#L225

https://github.com/cornell-zhang/heterocl/blob/267d670bfd7ec14276a7ddd21103a6e6ed668e5b/tvm/src/pass/generate_reuse_buffer.cc#L234

The index expression is simplified by `Simplify` function. However, `Simplify` function does not fold expression "across" cast node. For example:
```
(int33) (y + 1) - (int33) y
```
does not get simplified to `1`. We need to remove the cast operations and provide simplifier such expression: `y + 1 - 1`

The `CastRemover` in `tensor.py` is added for this purpose:
https://github.com/cornell-zhang/heterocl/blob/267d670bfd7ec14276a7ddd21103a6e6ed668e5b/python/heterocl/tensor.py#L151

However, removing all cast node for index expression leads to unsafe binary operation type. For example: 
https://github.com/zzzDavid/heterocl/blob/8875efb47a1978ce8b1064ebb271c76455c4a307/tests/issues/test_issue_386.py#L9-L17
When all the cast operations in the tensor index expression are removed, the `Mul` operation will have a mismatch for its two operand types.

This PR fixes this issue by removing the `CastRemover` in the frontend, so that this issue doesn't happen. But `ReuseBufferInserter` still faces the issue that simplifier doesn't work with cast operation. Therefore a `CastRemover` is added precisely where it's needed: right before when an index expression gets simplified. 

> Is it safe to remove the cast operations?
Seems the CastRemover does not enumerate all the operations but only several binary ones.

As explained above, CastRemover only removes cast in the index expression which needs to be simplified, so only binary operations and single variable needs to be covered since it's not meant to remove cast in all cases. 

Other than that, I also added type checking in the CastRemover, in case the binary operation cannot be cast-removed.  

> Can you also provide the test cases for reuse_at? Seems you changed the reuse buffer pass.

The changes are meant to make `reuse_at` work without removing all cast in tensor indices, instead of changing its behaviour. 

So the original tests in `test_schedule_memory.py` passing indicates the fix is working as expected.

zzzDavid commented:
@chhzh123 @Shaojie Please review again and merge if there's no issue

hecmay commented on tvm/src/pass/generate_reuse_buffer.cc:
```diff
@@ -55,6 +55,124 @@ class ModulusRemover final : public IRMutator {
+// Example: (cast(x) + cast(y)) -> (x + y)
+// Usage: CastRemover castRemover;
+//        Expr expr = castRemover.Mutate(expr);
+class CastRemover final : public IRMutator {
```
I would suggest moving this CastRemover class to `ir_utils.cc`, so that it can also be used in some other passes.

hecmay commented:
@zzzDavid I've merged the fix into master. you can pull it back and it should pass the CI/CD tests

hecmay commented:
the zhang-x1 local CI/CD runner cannot run through

zzzDavid commented:
@hecmay Thanks, I'll take a look at it. It seems my local runner is not using the correct g++. Currently it works on your local runner. Please feel free to merge it.
